### PR TITLE
Introduce `onActionCallback` prop in Kedro-Viz react component

### PR DESCRIPTION
**Pull request by jitu5**
## Description

Resolves https://github.com/kedro-org/kedro-viz/issues/2022

This PR addresses a requirement for users of Kedro-Viz as a React component in embedded mode. It allows displaying node metadata in a custom UI. Additionally, it supports the VSCode extension use case where clicking on a node passes the node data to the extension to open the relevant file.

```
<KedroViz 
  data={data} 
  options={...}
  onActionCallback={(action) => {
    console.log('Action: ', action);
  }}
/>
```

Below is the action format, which will be pass as argument to call back function  
```
action = {
  type: <ACTION_NAME>,
  payload: <PAYLOAD_OF_ANY_TYPE>
}
```

Note: We can not put `onNodeClickCallback` under a `options` prop as `onActionCallback` is a callback function and options are Redux state, it should only contain plain JavaScript objects, arrays, and primitives.

## Development notes

- Callback is pass to store via `configureStore` in App.js
- Callback call happens in `createCallbackMiddleware`
- On `TOGGLE_NODE_CLICKED` action intercept in middleware, using clicked node id and `state.graph.nodes` entire node object is pass as an argument to callback.  

## QA notes

You can test new prop in src/components/container.js as <App /> is the entry point or top level component for a standalone use case.

```
import React from 'react';
import App from '../app';
import getPipelineData from '../../utils/data-source';
import './container.scss';

/**
 * Top-level component for the use-case where Kedro-Viz is run as a standalone
 * app rather than imported as a library/package into a larger application.
 */
const Container = () => (
  <>
    <App
      options={{
        display: {
          globalNavigation: false,
          metadataPanel: false,
          sidebar: false,
        },
        theme: 'dark',
      }}
      onActionCallback ={(action) => {
        console.log('Action: ', action);
      }}
      data={getPipelineData()}
    />
  </>
);

export default Container;

```

## Checklist

- [X] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [ ] Added tests to cover my changes

### CODE REVIEW

File: src/store/middleware.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+const createCallbackMiddleware =
```
Comment (Huongg) [inline src/store/middleware.js:1]: hey @jitu5 the code looks great, if you can add some context as a comment here would be great.

**Review by Huongg — COMMENTED** [review]
hey generally the PR is looking pretty good and simple, thank you @jitu5 . My first thought would be what happens if we introduce another callback? eg: for slicing, run command etc. Would it be better to have a group for all the callback? Let me know your thought 

```
<KedroViz 
  data={data} 
  options={...}
  callbacks={
     onNodeClick={(node) => {
         console.log('Node clicked:', node);
      }} 
     onRunCommandClick={(runCommand) => ...}
  }
/>
```

**Comment by jitu5** [comment]
> hey generally the PR is looking pretty good and simple, thank you @jitu5 . My first thought would be what happens if we introduce another callback? eg: for slicing, run command etc. Would it be better to have a group for all the callback? Let me know your thought
> 
> ```
> <KedroViz 
>   data={data} 
>   options={...}
>   callbacks={
>      onNodeClick={(node) => {
>          console.log('Node clicked:', node);
>       }} 
>      onRunCommandClick={(runCommand) => ...}
>   }
> />
> ```


@Huongg I made changes to accept any event or action if we want to use a callback, as shown below.

```
<KedroViz 
  data={data} 
  options={...}
  onActionCallback={(action) => {
    console.log('Action: ', action);
  }}
/>
``` 


Below is the action format, which will be pass as argument to call back function
```
action = {
  type: <ACTION_NAME>,
  payload: <PAYLOAD_OF_ANY_TYPE>
}
```

File: src/store/middleware.js (hunk 1 above)
Comment (jitu5) [reply src/store/middleware.js:13]: Added.

**Review by rashidakanchwala — APPROVED** [review]
Look's great!. thanks @jitu5

**Review by Huongg — APPROVED** [review]
thanks for the update @jitu5 . looks good to me too

**Pull request merged**